driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -0,0 +1,1051 @@
+                retryState, operationContext,
+                () -> withSourceAndConnection(binding::getWriteConnectionSource, true, (connectionSource, connection) -> {
+                    ConnectionDescription connectionDescription = connection.getDescription();
+                    boolean effectiveRetryWrites = isRetryableWrite(retryWritesSetting, effectiveWriteConcern, connectionDescription, sessionContext);
+                    retryState.breakAndThrowIfRetryAnd(() -> !effectiveRetryWrites);
+                    resultAccumulator.onNewServerAddress(connectionDescription.getServerAddress());
+                    retryState.attach(AttachmentKeys.maxWireVersion(), connectionDescription.getMaxWireVersion(), true)
+                            .attach(AttachmentKeys.commandDescriptionSupplier(), () -> BULK_WRITE_COMMAND_NAME, false);
+                    BsonDocumentWrapper<?> lazilyEncodedBulkWriteCommand = createBulkWriteCommand(
+                            effectiveRetryWrites, effectiveWriteConcern, sessionContext, unexecutedModels, batchEncoder,
+                            () -> retryState.attach(AttachmentKeys.retryableCommandFlag(), true, true));
+                    return executeBulkWriteCommandAndExhaustOkResponse(
+                            retryState, connectionSource, connection, lazilyEncodedBulkWriteCommand, unexecutedModels,
+                            effectiveWriteConcern, operationContext);
+                })
+        );
+        try {
+            ExhaustiveBulkWriteCommandOkResponse bulkWriteCommandOkResponse = retryingBatchExecutor.get();
+            return resultAccumulator.onBulkWriteCommandOkResponseOrNoResponse(
+                    batchStartModelIndex, bulkWriteCommandOkResponse, batchEncoder.intoEncodedBatchInfo());
+        } catch (MongoWriteConcernWithResponseException mongoWriteConcernWithOkResponseException) {
+            return resultAccumulator.onBulkWriteCommandOkResponseWithWriteConcernError(
+                    batchStartModelIndex, mongoWriteConcernWithOkResponseException, batchEncoder.intoEncodedBatchInfo());
+        } catch (MongoCommandException bulkWriteCommandException) {
+            resultAccumulator.onBulkWriteCommandErrorResponse(bulkWriteCommandException);
+            throw bulkWriteCommandException;
+        } catch (MongoException e) {
+            // The server does not have a chance to add "RetryableWriteError" label to `e`,
+            // and if it is the last attempt failure, `RetryingSyncSupplier` also may not have a chance
+            // to add the label. So we do that explicitly.
+            shouldAttemptToRetryWriteAndAddRetryableLabel(retryState, e, false);
+            resultAccumulator.onBulkWriteCommandErrorWithoutResponse(e);
+            throw e;
+        }
+    }
+
+    /**
+     * @throws MongoWriteConcernWithResponseException This internal exception must be handled to avoid it being observed by an application.
```

Comment:
Please remind me why we have to both throw and catch this weird exception within even the same class?

Reply:
1) Our retry logic is built around exceptions: they are used to signal that an attempt has failed, to carry the error codes and the retryable label. A write concern error is, however, a weird one:

2) `Connection.command` does not throw when a write concern error happens.
3) There may be a useful server response accompanying the write concern error.

`1)` and `2)` mean that we must throw an "artificial" exception when a write concern error happens, if we want retries. `3)` means that we also must preserve the result when doing so, as it becomes useful if our "artificial" exception turns out to be the failure chosen as the outcome of all attempts. That's why we attach the response to the exception. And given that the exception is "artificial", we must ensure that once the retrying logic has done its work, we destroy the exception and never expose it to an application.